Improve documentation (Closes #14 and Closes #15) (#17)
gnosis/dao-module: docs/setup_guide.md

> # Setup Guide

Reviewer: I would call this section `SafeSnap Setup Guide`

> This guide will provide the steps how to setup a DAO module with a Safe on Rinkeby. The setup will use Realitio and can be used with Snapshot spaces.

Reviewer: ~This guide will provide the steps how to setup a DAO module with a Safe on Rinkeby.~
This guide shows how to setup the DAO module with a Gnosis Safe on the Rinkeby testnetwork.

Add link to https://realit.io/ and to https://snapshot.org/?

> More information how this works can be found on the [Gnosis blog](https://blog.gnosis.pm/ea67eb95c34f).

Reviewer: ~More information how this works can be found on the~
For more information on SafeSnap please refer to the [Gnosis blog](link).

> For more information on SafeSnap please refer to the [Gnosis blog](https://blog.gnosis.pm/ea67eb95c34f).

Reviewer 2: I understand `SafeSnap` to be the name for the entire concept/infrastructure (frontend/backend/contract) allowing users in a DAO to execute successful proposals onchain. I think this should be stated somewhere in the main readme, since I came here with the intention of `Follow our [setup guide](./docs/setup_guide.md) to get started with the DAO module` but I got a guide that explains how to set up a mysterious `SafeSnap`.

Reviewer 3: Yes, I can see why the naming is a little confusing.
SafeSnap is the name of the specific configuration using a Gnosis Safe, the dao module, and the corresponding Snapshot plugin. We should probably make that more explicit in the link from the readme.

Author: I adjusted the link in the readme to make this more clear

## Prerequisites

> To start the process you need to create a Safe (e.g. via https://app.gnosis-safe.io). This Safe will represent the DAO and hold all the assets (e.g. tokens and collectibles). It is required that you can execute via the Safe to setup the DAO module.

Reviewer: Shouldn't this be the Rinkeby link? rinkeby.gnosis-safe.io/app/

~It is required that you can execute via the Safe to setup the DAO module.~
A Safe transaction is required to setup the DAO module.

For the hardhat tasks to work the environment needs to be properly configured. See the [sample env file](../.env.sample) for more information.

The guide will use the Rinkeby ETH Realitio contract at [`0x3D00D77ee771405628a4bA4913175EcC095538da`](https://rinkeby.etherscan.io/address/0x3D00D77ee771405628a4bA4913175EcC095538da#code). Other network addresses can be found in the truffle build folder on the [Realitio GitHub repo](https://github.com/realitio/realitio-contracts). E.g. on mainnet the ETH Realitio contract can be found at [`0x325a2e0f3cca2ddbaebb4dfc38df8d19ca165b47`](https://etherscan.io/address/0x325a2e0f3cca2ddbaebb4dfc38df8d19ca165b47#code).

DISCLAIMER: Check the deployed Realitio contracts before using them.

## Setting up the module

> The first step is to deploy the module. Every DAO will have their own module. The module is linked to a DAO (called executor in the contract) and a oracle (e.g. Realitio). These cannot be changed after deployment.

Reviewer: a**n** oracle

> As part of the setup you need to define/ choose a template on Realitio. More information can be found in [their docs](https://github.com/realitio/realitio-dapp#structuring-and-fetching-information) 

Reviewer: ~define/ choose~
define or choose

### Setup the Realitio template

To define your own template a hardhat task is provided in the repository. It is possible to provide a template to that task via `--template` else the [default template](../src/tasks/defaultTemplate.json) is used.

The template should have the following format:
```json
{
    "title": "Did the proposal with the id %s pass the execution of the transactions with hash 0x%s?",
    "lang": "en",
    "type": "bool",
    "category": "DAO proposal"
}
```

- It is important that the `type` is `bool` as the module expects the outcome reported by Realitio to be `true`, `false` or `INVALID`
- The `category` and `lang` can be freely choosen and are only used in the Realitio web interfaces
- The title will also be displayed in the Realitio web interface and MUST include two `%s` placeholders
  - The first placeholder is for the `id` of the proposal (e.g. a ipfs hash)
  - The second placeholder is the hash of the concatenation of the EIP-712 transaction hashes (see the [README](../README.md) for more information)

> - IMPORTANT: The template should make it clear when to vote how on your questions

Reviewer 2: ```suggestion
- IMPORTANT: The template should make it clear when and how to vote on your questions
```

  - An example can be found in the [🍯DAO requirements](https://cloudflare-ipfs.com/ipfs/QmeJwtwdG4mPzC8sESrW7zqixZqdHDYnREz6ar9GCewgz7/)
  - DISCLAIMER: DO NOT BLINDLY COPY THE REQUIREMENTS. You should check the requirements and make the adjustments for your setup.

Using this template you can run the task by using `yarn hardhat --network <network> createDaoTemplate --oracle <oracle address> --template <your template json>` and this should provide you with a template id.

An example for this on Rinkeby would be (using the default template):
`yarn hardhat --network rinkeby createDaoTemplate ---oracle 0x3D00D77ee771405628a4bA4913175EcC095538da`

For this guide we will assume that the returned template id is `0x0000000000000000000000000000000000000000000000000000000000000dad`

### Deploying the module

> Now that we have a template the setup hardhat task can be used to deploy a DAO module instance. The setup task requires the following parameters `dao` (the address of the Safe), `oracle` (the address of the Realitio contract) and `template` (the template to used with Realitio). There are also optional parameters for more information run `yarn hardhat setup --help`.

Reviewer 2: The sentence was a bit difficult to follow:
```suggestion
Now that we have a template, a hardhat task can be used to deploy a DAO module instance. This setup task requires the following parameters: `dao` (the address of the Safe), `oracle` (the address of the Realitio contract) and `template` (the template to be used with Realitio). There are also optional parameters, for more information run `yarn hardhat setup --help`.
```

An example for this on Rinkeby would be:
`yarn hardhat --network rinkeby setup --dao <safe_address> --oracle 0x3D00D77ee771405628a4bA4913175EcC095538da --template 0x0000000000000000000000000000000000000000000000000000000000000dad`

This should return the address of the deployed DAO module. For this guide we assume this to be `0x4242424242424242424242424242424242424242`

Once the module is deployed you should verify the source code. If you use a network that is Etherscan compatible and you configure the `ETHERSCAN_API_KEY` in your environment you can use the provided hardhat task to do this. 

An example for this on Rinkeby would be:
`yarn hardhat --network rinkeby verifyEtherscan --module 0x4242424242424242424242424242424242424242 --dao <safe_address> --oracle 0x3D00D77ee771405628a4bA4913175EcC095538da --template 0x0000000000000000000000000000000000000000000000000000000000000dad`

### Enabling the module

> To allow the DAO module to actually execute transaction it is required to enable it on the Safe that it is connected to. For this it is possible to use the Transaction Builder on https://app.gnosis-safe.io. For this you can follow our tutorial on [adding a module](https://help.gnosis-safe.io/en/articles/4934427-add-a-module).

Reviewer: I think this should be the rinkeby link as well define/ choose

## Snapshot integration

> Once the module is setup it is possible to configure a space on [Snapshot](https://snapshot.org/) to enable the DAO module plugin. For this the space configuration needs to include `"plugins": { "daoModule": { "address": "<module_address"} }`. An example for this can be found in the [🍯DAO space configuration](https://cloudflare-ipfs.com/ipfs/QmahDCSkdED9BLZ3VtH6aJ8P5TmvMYEfA7fJa4hGsvEpi2/).

Reviewer 2: Missing close angle bracket:

```suggestion
Once the module is setup it is possible to configure a space on [Snapshot](https://snapshot.org/) to enable the DAO module plugin. For this the space configuration needs to include `"plugins": { "daoModule": { "address": "<module_address>"} }`. An example for this can be found in the [🍯DAO space configuration](https://cloudflare-ipfs.com/ipfs/QmahDCSkdED9BLZ3VtH6aJ8P5TmvMYEfA7fJa4hGsvEpi2/).
```

Once your space is configured you can attach transactions to you proposals via the plugin section:

1. Open the plugin selection

![Open the plugin selection](./snapshot_plugin_section.png)

2. Add DAO module plugin

![Add DAO module plugin](./snapshot_add_plugin.png)

3. Add DAO module transaction

<img src="./snapshot_module_add_tx.png"
     alt="Add DAO module transaction"
     width="250"/>
<img src="./snapshot_module_tx_details.png"
     alt="Enter transactiond etails"
     width="250" />
<img src="./snapshot_module_tx_confirm.png"
     alt="Check transaction details"
     width="250"/>

4. Check preview of transactions

![Transactions preview](./snapshot_plugin_preview.png)

> Once the proposal has been resolved it is possible to use the plugin to submit the proposal to the DAO module via the plugin. 

Reviewer 2: The plugin is the same, right?
```suggestion
Once the proposal has been resolved it is possible to submit the proposal to the DAO module via the newly added plugin. 
```

Author: I would leave out `newly added` as this step normally happens after the proposal has been resolved (so 7 days later :D)

This can also be done via the hardhat tasks provided in this repository. For more information run `yarn hardhat addProposal --help` or `yarn hardhat executeProposal --help`.

Once the question is available it can be answered via the Realitio web interface (e.g. https://reality.eth.link/app/).

## Monitoring your module

As anyone can submit proposals to your module it is recommended to setup some monitoring. The DAO module relies on the oracle (e.g. Realitio) to provide the correct answer so that no malicious transactions are executed. In the worst case the executor (e.g. the connected Safe) can invalidate a submitted proposal (see [README](../README.md) for more information). 

To make sure that all the involved stakeholders can react in a timely manner, the events emitted by the module contract should be monitored. Each time a new proposal is submitted the contract will emit a `ProposalQuestionCreated` event with the following parameters:
```
event ProposalQuestionCreated(
    bytes32 indexed questionId, // e.g. Realitio question id
    string indexed proposalId // e.g. Snapshot proposal id
);
```

> There are different services available for this. E.g. the [OpenZepplin Defender Sentinel](https://docs.openzeppelin.com/defender/sentinel).

Reviewer: ~There are different services available for this. E.g. the~
There are different services available for this such as the
